content/openusercss-integration.js

```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+(() => {
+  const manifest = chrome.runtime.getManifest();
+
+  // Tell the page that we exist and that it should send the handshake
+  window.postMessage({
+    'type': 'ouc-begin-handshake'
+  }, '*');
```

Comment:
I think you should specify `location.origin` here.

Reply:
I don't see the way that would be useful or necessary. Can you elaborate what you mean?

Reply:
Just for the sake of completeness because this is how postMessage is supposed to be used. Theoretically though, another extension could open this tab and send a message from another tab :-)

Reply:
Actually, I would prefer a dedicated custom DOM event. Why are you using postMessage? It's a busy one.

Reply:
I pushed an origin whitelist, which covers for this and the event origin.
I'm using postMessage because when I use `CustomEvent`s, they don't trigger the listener in the content script. There's probably some sort of scoping, but checking for the `type` property filters out everything we don't need.

Reply:
Stylish/Stylus have always been using custom DOM events on USO so it might be better if you investigate why it doesn't trigger here.

Reply:
After looking into it, I think it's because the content security policy the site uses (https://www.w3.org/TR/CSP3/#should-block-inline).
I think it wouldn't be a good idea allow inline event handlers from just anywhere. I believe sendMessage uses the same message bus (on `window`) under the hood, so I'm pretty sure there's no significant drawback in doing it the current way.

Reply:
Stylus doesn't depend on inline event listeners because we bind them in js as you can see in [install-hook-userstyles.js](https://github.com/openstyles/stylus/blob/master/content/install-hook-userstyles.js#L14). BTW, I would bind them on `window` object so that they don't pollute the devtools elements inspector.

Reply:
Sorry, I think I used the wrong terminology here. What I wanted to mean is that listeners not added by a script that exists in the HTML head are not allowed to listen for custom events.  

All event listeners are bound to `window`.
In this PR: https://github.com/openstyles/stylus/pull/328/commits/6657ab53647e86dab23791aabc02c358ce941015#diff-057c3ffbe917f05ee2f42ce8d36884a9R18
On the site: [OpenUserCSS/openusercss.org/blob/extension-integration/src/client/js/browser.js#L93](https://github.com/OpenUserCSS/openusercss.org/blob/extension-integration/src/client/js/browser.js#L93)

Reply:
1. Scope of custom DOM events propagation isn't limited by CSP as far as I know, at least I don't see anything about that in the specification.
2. `message` event is always posted to the `window`, but I'm talking about custom DOM events

Reply:
Here's an example:

![pic](https://puu.sh/yVrPI/4e4e8f44e2.png)

1. the first line is executed in the extension's content script (using the toolbar selector)
2. the second line is executed here in github page context (using the toolbar selector)

As you can see the message is successfully delivered. The only complication is FF needs [`cloneInto`](https://github.com/openstyles/stylus/blob/master/content/install-hook-userstyles.js#L94).

Reply:
Well, come to think of it, we can use `message` on `window` since it works already.

Reply:
Awesome!
Implementing CustomEvents down the line could be useful though, I think it's definitely solvable.